applications/GeoMechanicsApplication/custom_processes/apply_c_phi_reduction_process.hpp

```diff
@@ -0,0 +1,187 @@
+            << "Try a c-phi reduction factor " << mReductionFactor << " (safety factor "
+            << 1. / mReductionFactor << ") Previous reduction = " << mPreviousReductionFactor
+            << " Reduction increment = " << mReductionIncrement << std::endl;
+
+        double    phi                = 0.;
+        double    reduced_phi        = 0.;
+        double    c                  = 0.;
+        double    reduced_c          = 0.;
+        IndexType previousPropertyId = -1;
+        // Apply C/Phi Reduction procedure for the model part:
+        block_for_each(mrModelPart.Elements(),
+                       [this, &phi, &reduced_phi, &c, &reduced_c, &previousPropertyId](Element& rElement) {
+            // Only compute new c and phi if the Id changes
+            if (rElement.GetProperties().Id() != previousPropertyId) {
+                phi                = GetAndCheckPhi(rElement.GetProperties());
+                reduced_phi        = ComputeReducedPhi(phi);
+                c                  = GetAndCheckC(rElement.GetProperties());
+                reduced_c          = mReductionFactor * c;
+                previousPropertyId = rElement.GetProperties().Id();
+            }
+            SetCPhiAtElement(rElement, reduced_phi, reduced_c);
+        });
+        KRATOS_CATCH("")
+    }
+
+    void ExecuteFinalizeSolutionStep() override
+    {
+        mPreviousReductionFactor = mReductionFactor;
+        KRATOS_INFO("ApplyCPhiReductionProcess::ExecuteFinalizeSolutionStep")
+            << "Reached safety factor sofar = " << 1.0 / mReductionFactor << std::endl;
+    }
+
+    void ExecuteFinalize() override
+    {
+        KRATOS_INFO("ApplyCPhiReductionProcess")
+            << "Final safety factor = " << 1.0 / mReductionFactor << std::endl;
+    }
+
+private:
+    ModelPart& mrModelPart;
+    double     mReductionFactor         = 1.0;
+    double     mPreviousReductionFactor = 1.0;
+    double     mReductionIncrement      = 0.1;
+
+    double GetAndCheckPhi(const Element::PropertiesType& rProp)
+    {
+        // Get the initial properties from the model part. Recall that we create a separate
+        // properties object with reduced c and phi for each and every element. Those reduced
+        // properties objects are not linked to the original ones.
+        const auto& part_properties = mrModelPart.GetProperties(rProp.Id());
+
+        // Check for UMAT PHI Parameter
+        KRATOS_ERROR_IF(!part_properties.Has(INDEX_OF_UMAT_PHI_PARAMETER) ||
+                        !part_properties.Has(NUMBER_OF_UMAT_PARAMETERS) || !part_properties.Has(UMAT_PARAMETERS))
+            << "Insufficient material data for C-Phi reduction process: " << std::endl;
+        KRATOS_ERROR_IF(part_properties[INDEX_OF_UMAT_PHI_PARAMETER] < 1 || part_properties[INDEX_OF_UMAT_PHI_PARAMETER] > part_properties[NUMBER_OF_UMAT_PARAMETERS])
+            << "undefined INDEX_OF_UMAT_PHI_PARAMETER: " << part_properties[INDEX_OF_UMAT_PHI_PARAMETER]
```

Comment:
Here, we'd probably better state that the index is invalid, since it is out of bounds:
```suggestion
            << "invalid INDEX_OF_UMAT_PHI_PARAMETER: " << part_properties[INDEX_OF_UMAT_PHI_PARAMETER] << " (out-of-bounds index)"
```

Reply:
changed the message